### PR TITLE
std: Stabilize APIs for the 1.7 release

### LINKED ISSUE #23284: Tracking issue for Path::relative_from stabilization
`Path::relative_from` removes the common prefix from a path, e.g. `/foo/bar` relative_from `/foo` yields `bar`.

The old `path_relative_from` would calculate the relative path from two points on the file system, e.g. the relative path from `/foo/bar` to `/foo/baz/qux` is `../baz/qux`. This is a useful operation.

Nominating because I think either `relative_from` is named wrong or is not useful.

### LINKED ISSUE #27712: Tracking issue for `fs_mode` stabilization
The `fs_mode` feature refers to platform-specific constants for file access modes exported in `std::os`. While these constants are OK as-is, we don't have a clear overall vision for which constants should be included in `std::os` or how they should be organized. Before we stabilize, we should establish such a vision.

cc @alexcrichton

### LINKED ISSUE #27722: Tracking issue for `path_prefix` stabilization
The `Path::prefix` method allows you to extract a `Prefix` from a path, if any -- which is relevant only on Windows, where the prefix gives e.g. volume information.

It's already possible to get this information through the `components` iterator, so it's not clear whether this additional convenience is worth exposing.

cc @retep998

### LINKED ISSUE #27728: Tracking issue for formatting with a radix
This is a tracking issue for the unstable `fmt_radix` feature in the standard library. These functions provide the ability to format a number in an arbitrary radix. Note that base 2, 8, and 10, and 16 are already supported via various formatting flags.

Needs a decision about whether this functionality should even be in the standard library, and it will perhaps want a different interface as well.

### LINKED ISSUE #27729: Tracking issue for `as_slice` stabilization
Over time, the way that you convert e.g. a vector to a slice has transformed repeatedly. With deref coercions, it's usually implicit, but there are still some cases where you need to do it explicitly.

The `as_slice` method follows the usual conversion conventions, but has been somewhat controversial due to offering "yet another way" to perform the conversion (in addition to `as_ref` and `&*`).

### LINKED ISSUE #27735: Tracking issue for `IntoCow` stabilization
The `IntoCow` trait is a custom conversion trait for working with `Cow`; the custom nature is needed to deal with trait coherence issues.

Ultimately, the hope was that this trait could support ergonomic "ownership polymorphism" for things like the `entry` API. That's not the case today due to coherence limitations, but may become possible with [impl specialization](https://github.com/rust-lang/rfcs/pull/1210) or other changes to the trait system.

The API needs a full re-evaluation before stabilization.

### LINKED ISSUE #27755: Tracking issue for `wrapping` stabilization
The `std::num::wrapping` module provides a trait, `OverflowingOps`, that permits overflowing arithmetic on primitive types and tells whether an overflow occurred.

Other methods, like `saturating_add` and `wrapping_add`, are included directly as inherent methods on these types.

It's not clear whether the `overflowing_foo` methods are useful, and if so, whether there's any reason for them to live on a trait. Ideally, if we kept them, they'd move onto the primitive types, and the `wrapping` module would be removed.

### LINKED ISSUE #27782: Tracking issue for DebugTuple::formatter
This is a tracking issue for the unstable feature `debug_builder_formatter` in the standard library. This method was add relatively recently, and my thoughts on it are:

* It seems odd to have this function only on `DebugTuple` and no other builders.
* It does seem reasonable to allow reading out the formatter for flags and such in various situations perhaps.

cc @sfackler

### LINKED ISSUE #27798: Tracking issue for sync::Semaphore
This is a tracking issue for the unstable `semaphore` feature of the standard library. 

* Should we have this kind of semaphore?
* Should we allow negative counts?
* Should there be an RAII guard for a "resource"?

Decision!

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit stabilizes and deprecates the FCP (final comment period) APIs for
the upcoming 1.7 beta release. The specific APIs which changed were:

Stabilized
- `Path::strip_prefix` (renamed from `relative_from`)
- `path::StripPrefixError` (new error type returned from `strip_prefix`)
- `Ipv4Addr::is_loopback`
- `Ipv4Addr::is_private`
- `Ipv4Addr::is_link_local`
- `Ipv4Addr::is_multicast`
- `Ipv4Addr::is_broadcast`
- `Ipv4Addr::is_documentation`
- `Ipv6Addr::is_unspecified`
- `Ipv6Addr::is_loopback`
- `Ipv6Addr::is_unique_local`
- `Ipv6Addr::is_multicast`
- `Vec::as_slice`
- `Vec::as_mut_slice`
- `String::as_str`
- `String::as_mut_str`
- `<[T]>::clone_from_slice` - the `usize` return value is removed
- `<[T]>::sort_by_key`
- `i32::checked_rem` (and other signed types)
- `i32::checked_neg` (and other signed types)
- `i32::checked_shl` (and other signed types)
- `i32::checked_shr` (and other signed types)
- `i32::saturating_mul` (and other signed types)
- `i32::overflowing_add` (and other signed types)
- `i32::overflowing_sub` (and other signed types)
- `i32::overflowing_mul` (and other signed types)
- `i32::overflowing_div` (and other signed types)
- `i32::overflowing_rem` (and other signed types)
- `i32::overflowing_neg` (and other signed types)
- `i32::overflowing_shl` (and other signed types)
- `i32::overflowing_shr` (and other signed types)
- `u32::checked_rem` (and other unsigned types)
- `u32::checked_shl` (and other unsigned types)
- `u32::saturating_mul` (and other unsigned types)
- `u32::overflowing_add` (and other unsigned types)
- `u32::overflowing_sub` (and other unsigned types)
- `u32::overflowing_mul` (and other unsigned types)
- `u32::overflowing_div` (and other unsigned types)
- `u32::overflowing_rem` (and other unsigned types)
- `u32::overflowing_neg` (and other unsigned types)
- `u32::overflowing_shl` (and other unsigned types)
- `u32::overflowing_shr` (and other unsigned types)
- `ffi::IntoStringError`
- `CString::into_string`
- `CString::into_bytes`
- `CString::into_bytes_with_nul`
- `From<CString> for Vec<u8>`
- `From<CString> for Vec<u8>`
- `IntoStringError::into_cstring`
- `IntoStringError::utf8_error`
- `Error for IntoStringError`

Deprecated
- `Path::relative_from` - renamed to `strip_prefix`
- `Path::prefix` - use `components().next()` instead
- `os::unix::fs` constants - moved to the `libc` crate
- `fmt::{radix, Radix, RadixFmt}` - not used enough to stabilize
- `IntoCow` - conflicts with `Into` and may come back later
- `i32::{BITS, BYTES}` (and other integers) - not pulling their weight
- `DebugTuple::formatter` - will be removed
- `sync::Semaphore` - not used enough and confused with system semaphores

Closes #23284
cc #27709 (still lots more methods though)
Closes #27712
Closes #27722
Closes #27728
Closes #27735
Closes #27729
Closes #27755
Closes #27782
Closes #27798

### CODE REVIEW

**Comment by alexcrichton** [comment]
r? @aturon 

Note that this PR currently leaves out the stabilization of [custom hashers in hashmaps](https://github.com/rust-lang/rust/issues/27713) because there's a bit of a question of where the `BuildHasher` trait goes that we didn't quite make an explicit decision on. I'd be fine backporting that stabilization, however.

File: src/libstd/net/ip.rs [hunk 1]
```diff
@@ -89,6 +89,9 @@ impl Ipv4Addr {
     }
 
     /// Returns true if this is a loopback address (127.0.0.0/8).
+    ///
+    /// This property is defined by RFC 6890
```
Comment (aturon) [inline src/libstd/net/ip.rs]: Should we disambiguate this RFC from our normal RFCs? Do we want to provide a link?

**Comment by aturon** [comment]
OK, r=me unless you want to deal with the tiny nit.

File: src/libstd/net/ip.rs (hunk 1 above)
Comment (alexcrichton) [inline src/libstd/net/ip.rs]: Nah I think that if you're reading this much into the documentation you'll probably recognize in the context that these are the other kind of RFCs. That being said if it ends up being confusing we can certainly update!

**Comment by alexcrichton** [comment]
oh also, cc @rust-lang/libs

**Comment by ollie27** [comment]
It looks like `checked_neg` is still missing for unsigned ints.

**Comment by alexcrichton** [comment]
@bors: r-

Oh that was originally left out because it always overflowed (e.g. always returned `None`), but now we special case 0 and it seems good for consistency, so I'll add.

File: src/libcore/num/mod.rs [hunk 2]
```diff
@@ -1557,6 +1527,26 @@ macro_rules! uint_impl {
             }
         }
 
+        /// Checked negation. Computes `!self`, returning `None` unless `self ==
```
Comment (eefriedman) [inline src/libcore/num/mod.rs]: `-self`?

Comment (alexcrichton) [inline src/libcore/num/mod.rs]: Oops, yes ideed!

File: src/libcore/slice.rs [hunk 3]
```diff
@@ -476,14 +475,12 @@ impl<T> SliceExt for [T] {
     }
 
     #[inline]
-    fn clone_from_slice(&mut self, src: &[T]) -> usize where T: Clone {
-        let min = cmp::min(self.len(), src.len());
-        let dst = &mut self[.. min];
-        let src = &src[.. min];
-        for i in 0..min {
-            dst[i].clone_from(&src[i]);
+    fn clone_from_slice(&mut self, src: &[T]) where T: Clone {
```
Comment (bluss) [inline src/libcore/slice.rs]: Oops, major perf regression. I'll try to PR this.

Comment (alexcrichton) [inline src/libcore/slice.rs]: Oh gah good eye!

(for those following along the PR [was submitted](https://github.com/rust-lang/rust/pull/31000))

Comment (bluss) [inline src/libcore/slice.rs]: Yep, it's fine. I'm not worried, as long as it gets into the same beta release.

**Comment by bluss** [comment]
Now I understand why the clone_from_slice fix got a p=1. `bytes::copy_memory` has been replaced by calls to `clone_from_slice` in many places.

I'm not even sure that's a good idea. `bytes::copy_memory` was a safe interface to memcpy, no matter what the compilation mode. Calling `clone_from_slice` will be slow in no optimizations mode, and will exercise the inliner and optimizer in optimizing mode.

**Pull request merged**